vispy/app/backends/_qt.py

```diff
@@ -563,50 +557,81 @@ def keyPressEvent(self, ev):
 
+    def _handle_native_gesture_event(self, ev):
+        if self._vispy_canvas is None:
+            return
+        t = ev.gestureType()
+        # this is a workaround for what looks like a Qt bug where
+        # QNativeGestureEvent gives the wrong local position.
+        # See: https://bugreports.qt.io/browse/QTBUG-59595
+        try:
+            pos = self.mapFromGlobal(ev.globalPosition())
+        except AttributeError:
+            # globalPos is deprecated in Qt6
+            pos = self.mapFromGlobal(ev.globalPos())
+        pos = pos.x(), pos.y()
+
+        if t == QtCore.Qt.NativeGestureType.BeginNativeGesture:
+            self._vispy_canvas.events.touch(
+                type='gesture_begin',
+                pos=_get_event_xy(ev),
+            )
+        elif t == QtCore.Qt.NativeGestureType.EndNativeGesture:
+            # TODO: need to track corresponding begin/end? not sure how in Qt5
+            self._native_touch_total_rotation = []
+            self._native_touch_total_scale = []
+            self._vispy_canvas.events.touch(
+                type='gesture_end',
+                pos=_get_event_xy(ev),
+            )
+        elif t == QtCore.Qt.NativeGestureType.RotateNativeGesture:
+            angle = ev.value()
+            last_angle = (
+                self._native_gesture_rotation_values[-1]
+                if self._native_gesture_rotation_values
+                else None
+            )
+            self._native_gesture_rotation_values.append(angle)
+            total_rotation_angle = math.fsum(self._native_gesture_rotation_values)
+            self._vispy_canvas.events.touch(
+                type="gesture_rotate",
+                pos=pos,
+                last_pos=None,  # TODO: is this needed?
+                rotation=angle,
+                last_rotation=last_angle,
+                total_rotation_angle=total_rotation_angle,
+            )
+        elif t == QtCore.Qt.NativeGestureType.ZoomNativeGesture:
+            scale = ev.value()
+            last_scale = (
+                self._native_gesture_scale_values[-1]
+                if self._native_gesture_scale_values
+                else None
+            )
+            self._native_gesture_scale_values.append(scale)
+            total_scale_factor = math.fsum(self._native_gesture_scale_values)
+            self._vispy_canvas.events.touch(
+                type="gesture_zoom",
+                pos=pos,
+                last_pos=None,  # TODO: is this needed?
+                last_scale=last_scale,
+                scale=scale,
+                total_scale_factor=total_scale_factor,
+            )
+        # TODO: do we need or will we ever get pan events?
+        # With two fingers they are anyway converted to scroll/wheel events.
+        # On macOS, more fingers are usually swallowed by the OS (by spaces,
+        # mission control, etc.).
+        # elif t == QtCore.Qt.NativeGestureType.PanNativeGesture:
```

Comment:
Unfortunately I can't (yet) get this event to trigger on macOS even after disabling a bunch of the OS-level trackpad gestures. The docs are thin but suggest this might only be supported in Wayland, and perhaps absent from Qt5.

Reply:
I still can't get this gesture to trigger on macOS. Then again I can't get the `PanGesture` with the previous method either. Qt converts the _two_ finger panning to `QWheelEvent` but you _can_ get the `DeviceType` to check if that's a touchpad or a mouse. This also seems a bit inconsistent though (in a VM it's always reporting as touchpad for me even when using my mouse).

I'm a little stuck testing on Linux as well. I only have a VM and I think it's doing some kind of device virtualization, so I don't get any gestures in the guest OS.